File: samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/infrastructure/zlink/spots/tictactoegamespot/handlers/PlayActorGetCurrentGameStateHandler.java
package systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.tictactoegamespot.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.framework.handlers.ZLinkSpotActorSend;
import systems.zlink.samples.tictactoe.server.configuration.SampleNames;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.actors.PlayActor;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.tictactoegamespot.TicTacToeGame;
import systems.zlink.samples.tictactoe.shared.contracts.JoinGameMsg;
import systems.zlink.samples.tictactoe.shared.contracts.JoinGameNotify;

import java.util.concurrent.CompletionStage;

@ZLinkHandlerGroup(SampleNames.PlayActor)
public final class PlayActorGetCurrentGameStateHandler {
    @ZLinkSpotActorSend
    public CompletionStage<Void> currentState(
            TicTacToeGame spot, PlayActor actor, ZLinkMessageContext context, JoinGameMsg request) {
        return actor.context()
                .boundSession()
                .send(new JoinGameNotify(spot.currentState(actor, request.roomId())))
                .submit();
    }
}

File: samples/kotlin/TicTacToe/Server/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/server/play/infrastructure/zlink/spots/tictactoegamespot/handlers/PlayActorGetCurrentGameStateHandler.kt
package systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.tictactoegamespot.handlers

import kotlinx.coroutines.future.await
import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.handlers.ZLinkSpotActorSend
import systems.zlink.samples.kotlin.tictactoe.server.configuration.SampleNames
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.actors.PlayActor
import systems.zlink.samples.kotlin.tictactoe.server.play.infrastructure.zlink.spots.tictactoegamespot.TicTacToeGame
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.JoinGameMsg
import systems.zlink.samples.kotlin.tictactoe.shared.contracts.JoinGameNotify

@ZLinkHandlerGroup(SampleNames.PlayActor)
class PlayActorGetCurrentGameStateHandler {
    @ZLinkSpotActorSend
    suspend fun currentState(
        spot: TicTacToeGame,
        actor: PlayActor,
        context: ZLinkMessageContext,
        request: JoinGameMsg,
    ) {
        actor
            .context()
            .boundSession()
            .send(JoinGameNotify(spot.currentState(actor, request.roomId)))
            .submit()
            .await()
    }
}
